relay-metrics/src/protocol.rs

```diff
@@ -152,20 +152,41 @@ fn is_valid_name(name: &str) -> bool {
     false
 }
 
-/// Validates an MRI of the form `<ty>:<ns>/<name>@<unit>`
-///
-/// Note that the format used in the statsd protocol is different: Metric names are not prefixed
-/// with `<ty>:` as the type is somewhere else in the protocol.
-pub(crate) fn is_valid_mri(name: &str) -> bool {
-    let mut components = name.splitn(2, ':');
-    if components.next().is_none() {
-        return false;
-    }
+/// A metric name parsed as MRI, a naming scheme which includes most of the metric's bucket key
+/// (excl. timestamp and tags).
+pub struct MetricMri {
+    /// The metric type.
+    pub ty: MetricType,
+    /// The namespace/usecase for this metric. For example `sessions` or `transactions`.
+    pub namespace: String,
+    /// The actual name, such as `duration` as part of `d:transactions/duration@ms`
+    pub name: String,
+    /// The metric unit.
+    pub unit: MetricUnit,
+}
+
+impl std::str::FromStr for MetricMri {
+    type Err = ParseMetricError;
 
-    if let Some(name_unit) = components.next() {
-        parse_name_unit(name_unit).is_some()
-    } else {
-        false
+    /// Parses and validates an MRI of the form `<ty>:<ns>/<name>@<unit>`
+    ///
+    /// Note that the format used in the statsd protocol is different: Metric names are not prefixed
+    /// with `<ty>:` as the type is somewhere else in the protocol.
+    fn from_str(name: &str) -> Result<Self, Self::Err> {
+        let (raw_ty, rest) = name.split_once(':').ok_or(ParseMetricError(()))?;
+        let ty = raw_ty.parse()?;
+
+        let (raw_namespace, rest) = rest.split_once('/').ok_or(ParseMetricError(()))?;
+        let namespace = raw_namespace.to_owned();
```

Comment:
Should `namespace` also become an `enum`, like `ty`? Then we could fail parsing here when the namespace is unknown, and match on the namespace enum in `send_metric_message`.

Reply:
If we do that we cannot forward unknown usecases from pops. I think there's some merit to that. Though I guess you could make the same argument for units. I don't know.

Reply:
I'm also strongly in favor of an enum for the namespace. See https://github.com/getsentry/relay/pull/1215 for a list of all namespaces that are initially defined.  Especially in the Kafka producer, this would remove the messy sequential check and instead become a plain `match metric.name.namespace` or better `metric.namespace()`.

You can still have an `Unknown` variant. See my comment on the `MetricMri` struct for more thoughts regarding forwarding.

> Though I guess you could make the same argument for units. I don't know.

We do actually forward unknown units explicitly, see: [`MetricUnit::Custom`](https://getsentry.github.io/relay/relay_metrics/enum.MetricUnit.html#variant.Custom).